experiment/scheduler.py

```diff
@@ -121,6 +140,436 @@ def end_expired_trials(experiment_config: dict):
+
+        # Bound for the number of nonpreemptibles we can start if the experiment
+        # specified preemptible_runners.
+        self.max_nonpreemptibles = min(
+            math.ceil(self.num_trials * self.NONPREEMPTIBLES_FRACTION),
+            self.MAX_NONPREEMPTIBLES)
+        logger.info('Max nonpreemptibles: %d.', self.max_nonpreemptibles)
+
+        # Bound for the number of preemptibles we can start if the experiment
+        # specified preemptible_runners.
+        self.max_preemptibles = (self.num_trials *
+                                 self.MAX_PREEMPTIBLES_MULTIPLIER)
+        logger.info('Max nonpreemptibles: %d.', self.max_preemptibles)
+
+        # Attributes for preemptible retry window. The preemptible retry window
+        # is a time period that starts when the last initial trial is started.
+        # It determines how long we can retry preempted trials using
+        # preemptibles. This bounds the length of time an experiment lasts.
+        self.preemptible_window = (experiment_config['max_total_time'] *
+                                   self.PREEMPTIBLE_WINDOW_MULTIPLIER)
+        self._initial_trials = list(
+            get_experiment_trials(experiment_config['experiment']))
+        self._max_time_started = None
+
+        self.preempted_trials = {}
+        self.preemptible_starts_futile = False
+
+        self.base_resource_url = (
+            'https://www.googleapis.com/compute/v1/projects/{project}/zones/'
+            '{zone}/instances/').format(
+                project=experiment_config['cloud_project'],
+                zone=experiment_config['cloud_compute_zone'])
+
+        # Filter operations happening before the experiment started.
+        self.last_preemptible_query = (db_utils.query(models.Experiment).filter(
+            models.Experiment.name == experiment_config['experiment']).one(
+            ).time_created.replace(tzinfo=datetime.timezone.utc))
+
+    def _get_max_time_started(self):
+        """Returns the last time_started of the self._initial_trials. Returns
+        None if any initial trials haven't been started yet. This is needed so
+        that the preemptible retry window starts from the end of the last
+        initial trial to be started."""
+        if self._max_time_started is not None:
+            return self._max_time_started
+
+        max_time_started = None
+        for trial in self._initial_trials:
+            time_started = trial.time_started
+            if time_started is None:
+                # An initial trial has never been started. Therefore the max
+                # time started doesn't exist and the window hasn't started.
+                return None
+
+            if max_time_started is None:
+                max_time_started = time_started
+                continue
+
+            max_time_started = max(time_started, max_time_started)
+
+        assert max_time_started is not None
+        max_time_started = max_time_started.replace(
```

Comment:
We do store them as utc, but we don't keep the timezone info in there because all times are utc in postgres. It's not clear to me how/if we can store the times as timezone aware: https://stackoverflow.com/questions/414952/sqlalchemy-datetime-timezone